honeybadgermpc/hbavss.py

```diff
@@ -229,57 +232,144 @@ def __exit__(self, typ, value, traceback):
-                # will be replaced by sending out IMPLICATE message later
-                # multicast(HbAVSSMessageType.IMPLICATE)
-                logging.error("PolyCommit verification failed.")
-                raise HoneyBadgerMPCError("PolyCommit verification failed.")
+                    commitments[k], self.my_id+1, shares[k], auxes[k], witnesses[k]):
+                # invalid share found and send all implicate msg
+                multicast((HbAVSSMessageType.IMPLICATE, self.private_key, k))
 
-        multicast(HbAVSSMessageType.OK)
+        multicast((HbAVSSMessageType.OK, ""))
 
-        # Bracha-style agreement
-        # to handle the case that a (malicious) node sends
-        # an OK or READY message multiple times
         ok_set = set()
         ready_set = set()
-        # if 2t+1 OK or t+1 READY -> send all ready
-        while len(ok_set) < 2 * self.t + 1 and len(ready_set) < self.t + 1:
-            sender, avss_msg = await recv()  # First value is the `sid`
-            if avss_msg == HbAVSSMessageType.OK and sender not in ok_set:
-                ok_set.add(sender)
-            elif avss_msg == HbAVSSMessageType.READY and sender not in ready_set:
-                ready_set.add(sender)
-
-        multicast(HbAVSSMessageType.READY)
+        r1_set = set()
+        r2_set = set()
+        ready_sent = False
+        r1_sent = False
+        r2_sent = False
+        r1_phi = [None] * self.n
+        r2_phi = [None] * self.n
 
-        # if 2t+1 ready -> output shares
-        while len(ready_set) < 2 * self.t + 1:
-            sender, avss_msg = await recv()  # First value is the `sid`
-            if avss_msg == HbAVSSMessageType.READY and sender not in ready_set:
+        while True:
+            # main recv loop for Bracha-style agreement and implicate handling
+            sender, avss_msg = await recv()
+            # OK
+            if avss_msg[0] == HbAVSSMessageType.OK and sender not in ok_set:
+                ok_set.add(sender)
+            # READY
+            if avss_msg[0] == HbAVSSMessageType.READY and sender not in ready_set:
                 ready_set.add(sender)
-
-        return shares
+            # IMPLICATE
+            if avss_msg[0] == HbAVSSMessageType.IMPLICATE:
+                # Handling Implication
+                j = sender
+                j_pk = avss_msg[1]
+                j_k = avss_msg[2]
+                # discard if PKj ! = g^SKj
+                if self.public_keys[j] != pow(self.g, j_pk):
+                    logger.debug("[%d] invalid implicate from sender %s",
+                                 self.my_id, sender)
+                    continue
+                # decrypt and verify
+                implicate_msg = await avid.retrieve(tag, j)
+                j_z = loads(implicate_msg)
+                j_shared_key = pow(ephemeral_public_key, j_pk)
+                j_share, j_aux, j_witnesses = SymmetricCrypto.decrypt(
+                    str(j_shared_key).encode("utf-8"), j_z[j_k])
```

Comment:
calling `encode()` is faster than calling `encode("utf-8")`

Reply:
Yeah, but we also encode it this way in the light version.

Reply:
You can change it both places. :)